### PR TITLE
cache proxy service instances to improve performance

### PR DESCRIPTION
**Pull request by hzxuzhonghu**

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -641,9 +641,12 @@ func (s *DiscoveryServer) initConnectionNode(discReq *xdsapi.DiscoveryRequest, c
 	}
 	con.mu.Unlock()
 
+	if err := con.modelNode.SetServiceInstances(s.Env.PushContext); err != nil {
+		return err
+	}
 	// Set the sidecarScope associated with this proxy if its a sidecar.
 	if con.modelNode.Type == model.SidecarProxy {
-		s.globalPushContext().SetSidecarScope(con.modelNode)
+		con.modelNode.SetSidecarScope(s.Env.PushContext)
```
Comment (ymesika) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Maybe use the `s.globalPushContext()` to get better protection?

File: pilot/pkg/model/context.go [hunk 2]
```diff
@@ -174,6 +177,33 @@ func (node *Proxy) GetRouterMode() RouterMode {
...
+// short cut here is useful only for CDS and parts of RDS generation code.
+//
+// Listener generation code will still use the SidecarScope object directly
+// as it needs the set of services for each listener port.
+func (node *Proxy) SetSidecarScope(ps *PushContext) {
+	instances := node.ServiceInstances
+	node.SidecarScope = ps.getSidecarScope(node, instances)
+}
+
+func (node *Proxy) SetServiceInstances(ps *PushContext) error {
```
Comment (ymesika) [inline pilot/pkg/model/context.go]: If what you need as an argument is `*Environment` why not have it?
Also can `ps` be nil here?

File: pilot/pkg/model/context.go [hunk 3]
```diff
@@ -174,6 +177,33 @@ func (node *Proxy) GetRouterMode() RouterMode {
...
+// proxy and updates the proxy Node. This is a convenience hack so that
+// callers can simply call push.Services(node) while the implementation of
+// push.Services can return the set of services from the proxyNode's
+// sidecar scope or from the push context's set of global services. Similar
+// logic applies to push.VirtualServices and push.DestinationRule. The
+// short cut here is useful only for CDS and parts of RDS generation code.
+//
+// Listener generation code will still use the SidecarScope object directly
+// as it needs the set of services for each listener port.
+func (node *Proxy) SetSidecarScope(ps *PushContext) {
```
Comment (ymesika) [inline pilot/pkg/model/context.go]: Can ps be nil?

Comment (rshriram) [reply pilot/pkg/model/context.go]: nope

File: pilot/pkg/model/context.go (hunk 2 above)
Comment (rshriram) [reply pilot/pkg/model/context.go]: nope..

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/ads.go]: make sense

**Comment by rshriram** [comment]
cc @wenchenglu

**Comment by wenchenglu** [comment]
@hzxuzhonghu could you create a github issue to track all these performance related work, and mark it as 1.1

**Comment by rshriram** [comment]
@wenchenglu this is related to https://github.com/istio/istio/issues/9296

**Comment by wenchenglu** [comment]
@rshriram thanks for adding the issue here

**Comment by costinm** [comment]
Do we have any numbers on how much this affects max connections or memory ?

I'm a bit worried - the service instances associated with a node may change ( and do change ) after connect. It seems the cache is updated when the node is pushed,  so likely all is ok ( it doesn't appear to be updated on incremental EDS - but again, it seems ok). We may have a problem even without the cache - not sure what's the best way to test and verify this is safe ( even without the cache - not entirely sure that an endpoint becoming a member of a service after connect was resulting in a push with the new inbound listeners).

**Comment by hzxuzhonghu** [comment]
>Do we have any numbers on how much this affects max connections or memory ?

No, we have no performance test env currently.

>the service instances associated with a node may change ( and do change ) after connect. 

Yes, correctly. It doesn't seem a problem. Because if the service instances associated with a proxy changes, pilot will begin a new full push.

**Pull request merged**